File: packages/python/src/output_guard/strategies/extract_json.py
"""extract-json: Extract first balanced JSON object/array from surrounding text.

Mirror of packages/ts/src/strategies/extract-json.ts.
"""

from __future__ import annotations


def extract_json(text: str) -> str:
    t = text.strip()
    if t.startswith("{") or t.startswith("["):
        return text
    obj_idx = t.find("{")
    arr_idx = t.find("[")
    if obj_idx == -1 and arr_idx == -1:
        return text
    if obj_idx == -1:
        start_idx, open_ch, close_ch = arr_idx, "[", "]"
    elif arr_idx == -1:
        start_idx, open_ch, close_ch = obj_idx, "{", "}"
    elif arr_idx < obj_idx:
        start_idx, open_ch, close_ch = arr_idx, "[", "]"
    else:
        start_idx, open_ch, close_ch = obj_idx, "{", "}"

    depth = 0
    in_string = False
    escaped = False
    for i in range(start_idx, len(t)):
        ch = t[i]
        if escaped:
            escaped = False
            continue
        if ch == "\\":
            escaped = True
            continue
        if ch == '"':
            in_string = not in_string
            continue
        if in_string:
            continue
        if ch == open_ch:
            depth += 1
        elif ch == close_ch:
            depth -= 1
            if depth == 0:
                return t[start_idx : i + 1]
    return text

File: packages/ts/src/strategies/extract-json.ts
/**
 * extract-json: Extract the first balanced JSON object or array from surrounding text.
 * Runs after strip-fences.
 */
export const extractJson = (text: string): string => {
  const t = text.trim();
  // If it already starts with { or [, skip
  if (t.startsWith("{") || t.startsWith("[")) {
    return text;
  }
  // Find first { or [
  const objIdx = t.indexOf("{");
  const arrIdx = t.indexOf("[");
  let startIdx = -1;
  let openChar: "{" | "[" = "{";
  let closeChar: "}" | "]" = "}";

  if (objIdx === -1 && arrIdx === -1) return text;
  if (objIdx === -1) { startIdx = arrIdx; openChar = "["; closeChar = "]"; }
  else if (arrIdx === -1) { startIdx = objIdx; }
  else if (arrIdx < objIdx) { startIdx = arrIdx; openChar = "["; closeChar = "]"; }
  else { startIdx = objIdx; }

  let depth = 0;
  let inString = false;
  let escaped = false;
  for (let i = startIdx; i < t.length; i++) {
    const ch = t[i];
    if (escaped) { escaped = false; continue; }
    if (ch === "\\") { escaped = true; continue; }
    if (ch === '"') { inString = !inString; continue; }
    if (inString) continue;
    if (ch === openChar) depth++;
    else if (ch === closeChar) {
      depth--;
      if (depth === 0) {
        return t.slice(startIdx, i + 1);
      }
    }
  }
  return text;
};
